[Core][2/3] Make streaming generator public API #38784 (#41436)
ray-project/ray: doc/source/ray-core/tasks/generators.rst

In these cases, we can use a remote generator function that returns a *dynamic* number of values.
To use this feature, set ``num_returns="dynamic"`` in the ``@ray.remote`` decorator or the remote function's ``.options()``.

> Then, when invoking the remote function, Ray will return a *single* ``ObjectRef`` that will get populated with an ``DynamicObjectRefGenerator`` when the task completes.

Reviewer: ```suggestion
When you invoke the remote function, Ray returns a *single* ``ObjectRef`` that is populated with a ``DynamicObjectRefGenerator`` when the task completes.
```

Author: this file is not relevant anymore. wont fix

> The ``DynamicObjectRefGenerator`` can be used to iterate over a list of ``ObjectRefs`` containing the actual values returned by the task.

Reviewer: ```suggestion
You can use the ``DynamicObjectRefGenerator`` to iterate over a list of ``ObjectRefs`` containing the actual values returned by the task.
```

.. literalinclude:: ../doc_code/generator.py
    :language: python
    :start-after: __dynamic_generator_start__
    :end-before: __dynamic_generator_end__

> We can also pass the ``ObjectRef`` returned by a task with ``num_returns="dynamic"`` to another task. The task will receive the ``DynamicObjectRefGenerator``, which it can use to iterate over the task's return values. Similarly, you can also pass an ``ObjectRefGenerator`` as a task argument.

Reviewer: ```suggestion
You can also pass the ``ObjectRef`` returned by a task with ``num_returns="dynamic"`` to another task. The task receives the ``DynamicObjectRefGenerator`` and can use it to iterate over the task's return values. Similarly, you can pass an ``ObjectRefGenerator`` as a task argument.
```

If a generator function raises an exception before yielding all its values, the values that it already stored will still be accessible through their ``ObjectRefs``.
The remaining ``ObjectRefs`` will contain the raised exception.
This is true for both static and dynamic ``num_returns``.

> If the task was called with ``num_returns="dynamic"``, the exception will be stored as an additional final ``ObjectRef`` in the ``DynamicObjectRefGenerator``.

Reviewer: ```suggestion
If you call the task with ``num_returns="dynamic"``, Ray stores the exception as an additional final ``ObjectRef`` in the ``DynamicObjectRefGenerator``.
```
